### ISSUE TITLE
Segfault after error in libmp3lame.so

### ISSUE THREAD

**Issue by nuess0r** [savonet/liquidsoap#1996]
**Describe the bug**
After ~13 days liquidsoap died with a segfault caused by libmp3lame.so:
```
# dmesg -e
[Sep27 10:40] liquidsoap[758]: segfault at 6f899f ip 00007f76876a8b98 sp 00007f7667fd9d18 error 4 in libmp3lame.so.0.0.0[7f768769e000+30000]
[  +0.000028] Code: 11 0c 8a 48 83 c1 01 4d 39 c3 75 b6 c3 0f 1f 44 00 00 85 c9 7e 64 4d 63 c9 44 8d 59 ff 31 ff 31 c9 49 c1 e1 02 0f 1f 44 00 00 <f3> 0f 10 0c 3e f3 0f 10 2c 3a 0f 28 f2 0f 28 fb 49 89 c8 4c 01 cf
```

Systemd journal was logging while the liquidsoap log was empty after logrotate (see #1995 ):
```
# journalctl -xb
Sep 27 10:40:10 relay liquidsoap[415]: 2021/09/27 10:40:10 [decoder:2] Decoding "/media/streamdata/jazz_swing/jazz/Erroll Garner-The Way You Look Tonight.mp3" ended: Ffmpeg_decoder.End_of_file.
Sep 27 10:40:10 relay liquidsoap[415]: [mp3 @ 0x7f7651d64780] Estimating duration from bitrate, this may be inaccurate
Sep 27 10:40:10 relay liquidsoap[415]: [mp3 @ 0x7f76524692c0] Estimating duration from bitrate, this may be inaccurate
Sep 27 10:40:10 relay liquidsoap[415]: 2021/09/27 10:40:10 [playlist_0:3] Prepared "/media/streamdata/jazz_swing/jazz/allen_henry_red/Henry Red Allen-Shine On Your Shoes.mp3" (RID 0).
Sep 27 10:40:10 relay liquidsoap[415]: 2021/09/27 10:40:10 [decoder:3] No decoder available for "/media/streamdata/jazz_swing/jazz/peterson_oscar/ws_ftp.log"!
Sep 27 10:40:11 relay kernel: liquidsoap[758]: segfault at 6f899f ip 00007f76876a8b98 sp 00007f7667fd9d18 error 4 in libmp3lame.so.0.0.0[7f768769e000+30000]
Sep 27 10:40:11 relay kernel: Code: 11 0c 8a 48 83 c1 01 4d 39 c3 75 b6 c3 0f 1f 44 00 00 85 c9 7e 64 4d 63 c9 44 8d 59 ff 31 ff 31 c9 49 c1 e1 02 0f 1f 44 00 00 <f3> 0f 10 0c 3e f3 0f 10 2c 3a 0f 28>
Sep 27 10:40:11 relay systemd[1]: liquidsoap@radio.service: Main process exited, code=killed, status=11/SEGV

```

That there is no encoder found for a txt file is clear but it seams that this doesn't caused the segfault as then liquidsoap normally just skips this file and goes to the next in the playlist. Liquidsoap segfaults then 1 second after that without telling me, which mp3 file it tried to open...

**To Reproduce**
I don't know yet how to reproduce

**Expected behavior**
Not crashing.
Skipping to next file or something like this would be OK.

**Version details**
 - OS: Debian bullseye
 - Version Liquidsoap 2.0.0-rc1
 - ffmpeg version 4.4 Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 10 (Debian 10.2.1-6)
  configuration: --disable-decoder=amrnb --disable-decoder=libopenjpeg --disable-gnutls --disable-liblensfun --disable-libopencv --disable-podpages --disable-sndio --disable-stripping --enable-avfilter --enable-avresample --enable-gcrypt --enable-gpl --enable-ladspa --enable-libaom --enable-libaribb24 --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libdavs2 --enable-libdc1394 --enable-libfdk-aac --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libiec61883 --enable-libjack --enable-libkvazaar --enable-libmp3lame --enable-libmysofa --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librist --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libtesseract --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libwebp --enable-libx265 --enable-libxavs2 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-nonfree --enable-openal --enable-opencl --enable-opengl --enable-openssl --enable-postproc --enable-pthreads --enable-shared --enable-version3 --enable-vulkan --incdir=/usr/include/x86_64-linux-gnu --libdir=/usr/lib/x86_64-linux-gnu --prefix=/usr --toolchain=hardened --enable-frei0r --enable-chromaprint --enable-libx264 --enable-vaapi --enable-libmfx --enable-libvmaf --enable-libsvtav1 --enable-libilbc --disable-altivec --shlibdir=/usr/lib/x86_64-linux-gnu
  libavutil      56. 70.100 / 56. 70.100
  libavcodec     58.134.100 / 58.134.100
  libavformat    58. 76.100 / 58. 76.100
  libavdevice    58. 13.100 / 58. 13.100
  libavfilter     7.110.100 /  7.110.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  9.100 /  5.  9.100
  libswresample   3.  9.100 /  3.  9.100
  libpostproc    55.  9.100 / 55.  9.100

**Install method**
apt packages from https://github.com/savonet/liquidsoap/releases

ffmpeg is installed from the debmultimedia project to have aac encoding support.

**Comment by toots** [comment]
Hi!

Thanks for this report. Have you ever used `gdb`? If the issue can be reproduced, this would be a great help for us to get a stack trace using it.

**Comment by nuess0r** [comment]
Some weeks ago I upgraded to liquidsoap 2.0.0

Today it happened again:
```
Nov 14 15:57:25 relay liquidsoap[3504143]: 2021/11/14 15:57:25 [decoder:2] Decoding "/media/streamdata/jazz_swing/singers/Billie Holiday-Yours And Mine.mp3" ended: Ffmpeg_decoder.End_of_file.
Nov 14 15:57:25 relay liquidsoap[3504143]: [mp3 @ 0x7fad5ed37bc0] Estimating duration from bitrate, this may be inaccurate
Nov 14 15:57:25 relay liquidsoap[3504143]: [mp3 @ 0x7fad5ed37bc0] Estimating duration from bitrate, this may be inaccurate
Nov 14 15:57:25 relay liquidsoap[3504143]: 2021/11/14 15:57:25 [jazz_swing:3] Prepared "/media/streamdata/jazz_swing/jazz/Zoot Sims-My Silent Love.mp3" (RID 15).
Nov 14 15:57:25 relay liquidsoap[3504143]: 2021/11/14 15:57:25 [decoder:3] No decoder available for "/media/streamdata/jazz_swing/singers/ws_ftp.log"!
Nov 14 15:57:26 relay kernel: traps: liquidsoap[3505230] general protection fault ip:7fad92ccab98 sp:7fad76e22d18 error:0 in libmp3lame.so.0.0.0[7fad92cc0000+30000]
Nov 14 15:57:26 relay systemd[1]: liquidsoap@radio.service: Main process exited, code=killed, status=11/SEGV
░░ Subject: Unit process exited
```
What surprises me is, that it prepares one file (Zoot Sims-My Silent Love.mp3) but the next message is about a completely different file name (ws_ftp.log, yes I could delete them but they've never been a problem with them being present).

I checked that the prepared file is present. I could make it available to test if the file is corrupt or so.

Why could it be, that the prepared file is not played/passed to the decoder?

Why could it be, that it was not logged what happened with the prepared file and what other file was prepared instead?

**Comment by vitoyucepi** [comment]
I think I can confirm that `libmp3lame.so` can cause segfault.
I'm not sure why, but It's probably somehow related to mp3 encoding.
I think so because there where no errors after I switched from `output.icecast` with encoding to `output.dummy`.

I have to do some more testing.
I use `ubuntu:20.04` in docker with 2.0.0 release apt package.

Is it necessary to use some special release with debug symbols or ordinary package is fine too?

**Comment by toots** [comment]
Hi all,

It would be very useful to get a gdb trace of the segfault. This can be done like this:
```
gdb /path/to/liquidsoap
> r <args>
.... crashes
> thread apply all bt
```
This should tell us exactly which part of the code is causing the segfault.

If attaching gdb to a running instance is not practical, it should also be possible to enable core dumps. This can be a little tricky but it's really useful once setup. Here's some doc I found by doing a quick google search: https://medium.com/@sourabhedake/core-dumps-how-to-enable-them-73856a437711

**Comment by vitoyucepi** [comment]
@toots is it necessary to use the package with the `dbgsym` affix to produce useful core dump?

**Comment by toots** [comment]
The package will provide debugging symbols for liquidsoap. There does not seem to be a debug package for lame of ffmpeg in debian...

**Comment by vitoyucepi** [comment]
@toots I've found repos for `debian` and `ubuntu` here
- https://wiki.debian.org/HowToGetABacktrace
- https://wiki.ubuntu.com/Debug%20Symbol%20Packages

But my question is about liquidsoap itself.
There's no dbgsym package for `ubuntu` [here](https://github.com/savonet/liquidsoap/releases/tag/v2.0.0) do I have to use `debian` instead?

**Comment by toots** [comment]
Yes, it would be nice to have the backtrace from liquidsoap as well by installing this package. If there's not debug package for ubuntu, debian seems more appropriate indeed. Thanks for the effort!

**Comment by vitoyucepi** [comment]
@toots I'm trying to figure out how to deploy create debug environment to server with docker.

In local environment I used this command
```
gdb --eval-command "set pagination 0" --eval-command "thread apply all bt" --pid 1
```
Because I attached to another container and liquidsoap `pid` is `1`
And from the beginning I have two errors.

<details>
<summary>Error as start of service</summary>

```
gdb_1         | Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
gdb_1         | caml_page_table_lookup (addr=0x7fa9235a6f08) at memory.c:97
gdb_1         | 97      memory.c: No such file or directory.
gdb_1         | 
gdb_1         | Thread 2 (Thread 0x7fa924ad9700 (LWP 15) "SRT:GC"):
gdb_1         | #0  0x00007fa934223ad8 in pthread_cond_timedwait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
gdb_1         | #1  0x00007fa9342d0f60 in srt::sync::Condition::wait_for (this=this@entry=0x7fa9342f43c8 <CUDT::s_UDTUnited+328>, lock=..., rel_time=...) at ./srtcore/sync.h:305
gdb_1         | #2  0x00007fa93426cee9 in CUDTUnited::garbageCollect (p=0x7fa9342f4280 <CUDT::s_UDTUnited>) at ./srtcore/api.cpp:2833
gdb_1         | #3  0x00007fa93421cea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
gdb_1         | #4  0x00007fa930957def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6
gdb_1         | 
gdb_1         | Thread 1 (Thread 0x7fa925c05380 (LWP 1) "liquidsoap"):
gdb_1         | #0  caml_page_table_lookup (addr=0x7fa9235a6f08) at memory.c:97
gdb_1         | #1  0x000055fc2e7865d5 in do_compare_val (stk=stk@entry=0x7ffd0e160e50, v1=140364419329800, v2=140364418432064, total=total@entry=1) at compare.c:174
gdb_1         | #2  0x000055fc2e786a07 in compare_val (total=1, v2=<optimized out>, v1=<optimized out>) at compare.c:97
gdb_1         | #3  caml_compare (v1=<optimized out>, v2=<optimized out>) at compare.c:312
gdb_1         | #4  0x000055fc2e6af7ad in camlStdlib__list__assoc_371 () at list.ml:192
gdb_1         | #5  0x000055fc2e2e7fb2 in camlEvaluation__lookup_183 () at lang/evaluation.ml:38
gdb_1         | #6  0x000055fc2e2e9514 in camlEvaluation__ans_896 () at lang/evaluation.ml:206
gdb_1         | #7  0x000055fc2e2eaa2a in camlEvaluation__eval_toplevel_inner_2403 () at lang/evaluation.ml:386
gdb_1         | #8  0x000055fc2e21ca6e in camlRuntime__fun_2495 () at lang/runtime.ml:38
gdb_1         | #9  0x000055fc2e33a124 in camlTutils__finalize_167 () at tools/tutils.ml:100
gdb_1         | #10 0x000055fc2e21d859 in camlRuntime__report_643 () at lang/runtime.ml:191
gdb_1         | #11 0x000055fc2e21db68 in camlRuntime__from_lexbuf_inner_2618 () at lang/runtime.ml:212
gdb_1         | #12 0x000055fc2e21ddca in camlRuntime__from_file_778 () at lang/runtime.ml:224
gdb_1         | #13 0x000055fc2e21de63 in camlRuntime__load_libs_inner_2635 () at lang/runtime.ml:235
gdb_1         | #14 0x000055fc2e0f8093 in camlMain__fun_2046 () at main.ml:90
gdb_1         | #15 0x000055fc2e0f8105 in camlMain__eval_294 () at main.ml:110
gdb_1         | #16 0x000055fc2e0f829d in camlMain__force_302 () at main.ml:126
gdb_1         | #17 0x000055fc2e5de53b in camlDtools__Dtools_impl__exec_786 () at src/dtools_impl.ml:455
gdb_1         | #18 0x000055fc2e6af141 in camlStdlib__list__iter_261 () at list.ml:110
gdb_1         | #19 0x000055fc2e5de578 in camlDtools__Dtools_impl__exec_786 () at src/dtools_impl.ml:458
gdb_1         | #20 0x000055fc2e0f7f92 in camlRunner__entry () at main.ml:566
gdb_1         | #21 0x000055fc2e0edc49 in caml_program ()
gdb_1         | #22 0x000055fc2e79f640 in caml_start_program ()
gdb_1         | #23 0x000055fc2e77dc84 in caml_startup_common (argv=0x7ffd0e161378, pooling=<optimized out>, pooling@entry=0) at startup_nat.c:158
gdb_1         | #24 0x000055fc2e77dcfb in caml_startup_exn (argv=<optimized out>) at startup_nat.c:163
gdb_1         | #25 caml_startup (argv=<optimized out>) at startup_nat.c:168
gdb_1         | #26 caml_main (argv=<optimized out>) at startup_nat.c:175
gdb_1         | #27 0x000055fc2e0eaaec in main (argc=<optimized out>, argv=<optimized out>) at main.c:41

```

</details>

<details>
<summary>Error when track changes</summary>

```
0x00007f8bc9d73ba2 in sigtimedwait () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 10 (Thread 0x7f8bb27fc700 (LWP 27) "liquidsoap"):
#0  0x00007f8bc9e2a3ff in poll () from target:/lib/x86_64-linux-gnu/libc.so.6
#1  0x000055dff907c591 in caml_poll (_read=<optimized out>, _write=<optimized out>, _err=<optimized out>, _timeout=<optimized out>) at duppy_stubs.c:95
#2  0x000055dff8eed190 in camlDuppy__poll_125 () at src/duppy.ml:37
#3  0x000055dff8eee2dc in camlDuppy__f_521 () at src/duppy.ml:208
#4  0x000055dff8eededd in camlDuppy__process_511 () at src/duppy.ml:229
#5  0x000055dff8eeec23 in camlDuppy__f_606 () at src/duppy.ml:326
#6  0x000055dff8eee9b4 in camlDuppy__queue_inner_2535 () at src/duppy.ml:357
#7  0x000055dff8c536e7 in camlTutils__process_729 () at tools/tutils.ml:175
#8  0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#9  0x000055dff90b8640 in caml_start_program ()
#10 0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180348040, arg=<optimized out>, arg@entry=1) at callback.c:111
#11 0x000055dff908efe0 in caml_thread_start (arg=0x7f8bac004c70) at st_stubs.c:538
#12 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#13 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 9 (Thread 0x7f8bb2ffd700 (LWP 26) "liquidsoap"):
#0  0x00007f8bcd7007b2 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000055dff908fa3f in st_condvar_wait (m=0x55dffa645420, c=0x7f8b9c002b70) at st_posix.h:290
#2  caml_condition_wait (wcond=<optimized out>, wmut=<optimized out>) at st_stubs.c:869
#3  0x000055dff8eeed45 in camlDuppy__f_606 () at src/duppy.ml:341
#4  0x000055dff8eee9b4 in camlDuppy__queue_inner_2535 () at src/duppy.ml:357
#5  0x000055dff8c536e7 in camlTutils__process_729 () at tools/tutils.ml:175
#6  0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#7  0x000055dff90b8640 in caml_start_program ()
#8  0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180349416, arg=<optimized out>, arg@entry=1) at callback.c:111
#9  0x000055dff908efe0 in caml_thread_start (arg=0x7f8bac0a5080) at st_stubs.c:538
#10 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#11 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 8 (Thread 0x7f8bb37fe700 (LWP 25) "liquidsoap"):
#0  0x00007f8bcd7007b2 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000055dff908fa3f in st_condvar_wait (m=0x55dffa645420, c=0x7f8ba0002b70) at st_posix.h:290
#2  caml_condition_wait (wcond=<optimized out>, wmut=<optimized out>) at st_stubs.c:869
#3  0x000055dff8eeed45 in camlDuppy__f_606 () at src/duppy.ml:341
#4  0x000055dff8eee9b4 in camlDuppy__queue_inner_2535 () at src/duppy.ml:357
#5  0x000055dff8c536e7 in camlTutils__process_729 () at tools/tutils.ml:175
#6  0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#7  0x000055dff90b8640 in caml_start_program ()
#8  0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180350744, arg=<optimized out>, arg@entry=1) at callback.c:111
#9  0x000055dff908efe0 in caml_thread_start (arg=0x7f8bac0a3900) at st_stubs.c:538
#10 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#11 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 7 (Thread 0x7f8bb3fff700 (LWP 24) "liquidsoap"):
#0  0x00007f8bcd7007b2 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000055dff908fa3f in st_condvar_wait (m=0x55dffa645420, c=0x7f8ba4002b70) at st_posix.h:290
#2  caml_condition_wait (wcond=<optimized out>, wmut=<optimized out>) at st_stubs.c:869
#3  0x000055dff8eeed45 in camlDuppy__f_606 () at src/duppy.ml:341
#4  0x000055dff8eee9b4 in camlDuppy__queue_inner_2535 () at src/duppy.ml:357
#5  0x000055dff8c536e7 in camlTutils__process_729 () at tools/tutils.ml:175
#6  0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#7  0x000055dff90b8640 in caml_start_program ()
#8  0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180352016, arg=<optimized out>, arg@entry=1) at callback.c:111
#9  0x000055dff908efe0 in caml_thread_start (arg=0x7f8bac0ab8e0) at st_stubs.c:538
#10 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#11 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 6 (Thread 0x7f8bb8dc8700 (LWP 23) "liquidsoap"):
#0  0x00007f8bc9dfcc61 in clock_nanosleep () from target:/lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f8bc9e02443 in nanosleep () from target:/lib/x86_64-linux-gnu/libc.so.6
#2  0x000055dff90878cc in posix_time2_9_nanosleep (x68=<optimized out>, x67=<optimized out>) at posix_time2_generated_stubs.c:112
#3  0x000055dff8f242e6 in camlPosix_time2__Posix_time2_generated_stubs__fun_878 () at posix-time2/src/posix_time2_generated_stubs.ml:132
#4  0x000055dff8f26571 in camlPosix_time2__fun_2352 () at posix-time2/src/posix_time2.ml:268
#5  0x000055dff8f2e6a5 in camlErrno_unix__raise_on_errno_inner_1901 ()
#6  0x000055dff8f2e962 in camlErrno_unix__with_unix_exn_1764 ()
#7  0x000055dff8c1cab7 in camlClock__sleep_449 () at clock.ml:43
#8  0x000055dff8c1da77 in camlClock__loop_1139 () at clock.ml:245
#9  0x000055dff8c1bded in camlClock__fun_2166 () at clock.ml:269
#10 0x000055dff8c536e7 in camlTutils__process_729 () at tools/tutils.ml:175
#11 0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#12 0x000055dff90b8640 in caml_start_program ()
#13 0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180353552, arg=<optimized out>, arg@entry=1) at callback.c:111
#14 0x000055dff908efe0 in caml_thread_start (arg=0x7f8bac0726b0) at st_stubs.c:538
#15 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#16 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 5 (Thread 0x7f8bb95d1700 (LWP 22) "liquidsoap"):
#0  0x00007f8bcd7007b2 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000055dff908fa3f in st_condvar_wait (m=0x55dffa647170, c=0x55dffa646f30) at st_posix.h:290
#2  caml_condition_wait (wcond=<optimized out>, wmut=<optimized out>) at st_stubs.c:869
#3  0x000055dff8c54a28 in camlTutils__fun_2294 () at tools/tutils.ml:340
#4  0x000055dff8c53084 in camlTutils__mutexify_105 () at tools/tutils.ml:90
#5  0x000055dff8ef7908 in camlDtools__Dtools_impl__thread_892 () at src/dtools_impl.ml:497
#6  0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#7  0x000055dff90b8640 in caml_start_program ()
#8  0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180194648, arg=<optimized out>, arg@entry=1) at callback.c:111
#9  0x000055dff908efe0 in caml_thread_start (arg=0x55dffa634390) at st_stubs.c:538
#10 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#11 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 4 (Thread 0x7f8bb9dd2700 (LWP 21) "liquidsoap"):
#0  0x00007f8bc9e2c8b3 in select () from target:/lib/x86_64-linux-gnu/libc.so.6
#1  0x000055dff908e3b1 in caml_thread_tick (arg=<optimized out>) at st_posix.h:388
#2  0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#3  0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 3 (Thread 0x7f8bba5d3700 (LWP 20) "liquidsoap"):
#0  0x00007f8bcd7007b2 in pthread_cond_wait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000055dff908fa3f in st_condvar_wait (m=0x55dffa5c1790, c=0x55dffa5c17c0) at st_posix.h:290
#2  caml_condition_wait (wcond=<optimized out>, wmut=<optimized out>) at st_stubs.c:869
#3  0x000055dff8ef88d2 in camlDtools__Dtools_impl__fun_2762 () at src/dtools_impl.ml:734
#4  0x000055dff8ef86ca in camlDtools__Dtools_impl__mutexify_1402 () at src/dtools_impl.ml:702
#5  0x000055dff8ef886f in camlDtools__Dtools_impl__f_1422 () at src/dtools_impl.ml:730
#6  0x000055dff8fbba05 in camlThread__fun_314 () at thread.ml:41
#7  0x000055dff90b8640 in caml_start_program ()
#8  0x000055dff90ae97d in caml_callback_exn (closure=closure@entry=140238180194896, arg=<optimized out>, arg@entry=1) at callback.c:111
#9  0x000055dff908efe0 in caml_thread_start (arg=0x55dffa63dca0) at st_stubs.c:538
#10 0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#11 0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 2 (Thread 0x7f8bbdfb6700 (LWP 17) "SRT:GC"):
#0  0x00007f8bcd700ad8 in pthread_cond_timedwait@@GLIBC_2.3.2 () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f8bcd7adf60 in srt::sync::Condition::wait_for (this=this@entry=0x7f8bcd7d13c8 <CUDT::s_UDTUnited+328>, lock=..., rel_time=...) at ./srtcore/sync.h:305
#2  0x00007f8bcd749ee9 in CUDTUnited::garbageCollect (p=0x7f8bcd7d1280 <CUDT::s_UDTUnited>) at ./srtcore/api.cpp:2833
#3  0x00007f8bcd6f9ea7 in start_thread () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#4  0x00007f8bc9e34def in clone () from target:/lib/x86_64-linux-gnu/libc.so.6

Thread 1 (Thread 0x7f8bbf0e2380 (LWP 1) "liquidsoap"):
#0  0x00007f8bc9d73ba2 in sigtimedwait () from target:/lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f8bcd704c8c in sigwait () from target:/lib/x86_64-linux-gnu/libpthread.so.0
#2  0x000055dff908ebf1 in caml_wait_signal (sigs=<optimized out>) at st_posix.h:474
#3  0x000055dff8ef7638 in camlDtools__Dtools_impl__wait_signal_789 () at src/dtools_impl.ml:471
#4  0x000055dff8ef77e1 in camlDtools__Dtools_impl__main_885 () at src/dtools_impl.ml:507
#5  0x000055dff8ef7a3a in camlDtools__Dtools_impl__catch_900 () at src/dtools_impl.ml:512
#6  0x000055dff8a12e02 in camlMain__fun_2257 () at main.ml:558
#7  0x000055dff8ef753b in camlDtools__Dtools_impl__exec_786 () at src/dtools_impl.ml:455
#8  0x000055dff8fc8141 in camlStdlib__list__iter_261 () at list.ml:110
#9  0x000055dff8ef7578 in camlDtools__Dtools_impl__exec_786 () at src/dtools_impl.ml:458
#10 0x000055dff8fc8141 in camlStdlib__list__iter_261 () at list.ml:110
#11 0x000055dff8ef7578 in camlDtools__Dtools_impl__exec_786 () at src/dtools_impl.ml:458
#12 0x000055dff8a10f92 in camlRunner__entry () at main.ml:566
#13 0x000055dff8a06c49 in caml_program ()
#14 0x000055dff90b8640 in caml_start_program ()
#15 0x000055dff9096c84 in caml_startup_common (argv=0x7fff3a7ac5f8, pooling=<optimized out>, pooling@entry=0) at startup_nat.c:158
#16 0x000055dff9096cfb in caml_startup_exn (argv=<optimized out>) at startup_nat.c:163
#17 caml_startup (argv=<optimized out>) at startup_nat.c:168
#18 caml_main (argv=<optimized out>) at startup_nat.c:175
#19 0x000055dff8a03aec in main (argc=<optimized out>, argv=<optimized out>) at main.c:41
```

</details>

The first error occurs only once, but another one is happened every track change.
Is everything goes ok. Also how can I ignore it in non-interactive session?

Anyway I pushed my current progress with the issue in [special branch](https://gitlab.com/vito-containers/liquidsoap/-/merge_requests/52).
Also I didn't installed some of debug symbols
```
libz3-4-dbgsym
libgl1-mesa-dri-dbgsym
libllvm11-dbgsym
libmfx1-dbgsym
libopenmpt0-dbgsym
librsvg2-2-dbgsym
```
is it necessary to have them in final build?

**Comment by toots** [comment]
Hi,

Thanks so much for your efforts.

Adding more debug symbols is not the hardest task so we can put that aside until we see if we need them. If the segfault does not involve them then we won't need them.

Running the whole system inside a `gdb` session may trigger things that are irrelevant w.r.t. to an actual run. I think your best course of action would be to find how to enable core dumps, specially if you are running from within a container.

Here's how we do it in our CI:
```yaml
      - name: Enable core dump
        run: |
          ulimit -c unlimited
          mkdir -p /tmp/core
          chown -R opam /tmp/core
          echo '/tmp/core/core.%h.%e.%t' > /proc/sys/kernel/core_pattern
```

You can see if core dumps are enabled by doing:
```
sleep 10 & killall -ABRT sleep
```

**Comment by vitoyucepi** [comment]
Finally I recorded some core dumps.
I've uploaded them all to [git+lfs repo](https://gitlab.com/vitoyucepi/liquidsoap-core-dumps).
Yet there's one more kind of segfaults to catch, `Code: c1`.

**Comment by toots** [comment]
Thanks! Can you also add the liquidsoap binary that was used to generate them?

**Comment by vitoyucepi** [comment]
@toots 
[Container sources](https://gitlab.com/vito-containers/liquidsoap/-/tree/upstream_issue_1996/containers/custom) with `liquidsoap.deb`.
Result docker image: `registry.gitlab.com/vito-containers/liquidsoap/build:custom-baaba299`

**Comment by toots** [comment]
Thank you so much for this, what an awesome report and data..

Here's one faulty trace:
```gdb
Thread 1 (Thread 0x7f3f6dccf700 (LWP 19)):
#0  0x00007f3f81dc8b98 in lame_copy_inbuffer (gfc=gfc@entry=0x7f3f6803d600, l=l@entry=0x800, r=r@entry=0x3, nsamples=nsamples@entry=1764, pcm_type=pcm_type@entry=pcm_float_type, jump=jump@entry=1, s=s@entry=1) at lame.c:1865
#1  0x00007f3f81dc9199 in lame_encode_buffer_template (gfp=gfp@entry=0x7f3f6809c8c0, buffer_l=0x800, buffer_r=0x3, nsamples=nsamples@entry=1764, mp3buf=mp3buf@entry=0x7f3f6dca9ec0 "*\214", mp3buf_size=mp3buf_size@entry=147456, pcm_type=pcm_float_type, aa=1, norm=norm@entry=1) at lame.c:1895
#2  0x00007f3f81dc9f55 in lame_encode_buffer_float (gfp=gfp@entry=0x7f3f6809c8c0, pcm_l=<optimized out>, pcm_r=<optimized out>, nsamples=nsamples@entry=1764, mp3buf=mp3buf@entry=0x7f3f6dca9ec0 "*\214", mp3buf_size=mp3buf_size@entry=147456) at lame.c:1925
#3  0x00005601e5a30c31 in ocaml_lame_encode_buffer_float_ba (l=<optimized out>, _bufl=<optimized out>, _bufr=<optimized out>) at lame_stubs.c:267
#4  0x00005601e589d549 in camlLame__fun_514 ()
#5  0x00005601e54b8dea in camlLame_encoder__encode_568 () at encoder/lame_encoder.ml:237
#6  0x00005601e553bcb6 in camlOutput__output_chunks_1356 () at outputs/output.ml:239
#7  0x00005601e5539f20 in camlOutput__fun_1955 () at outputs/output.ml:170
#8  0x00005601e564e402 in camlClock__fun_2240 () at clock.ml:299
#9  0x00005601e5990218 in camlStdlib__list__fold_left_275 () at list.ml:121
#10 0x00005601e564c06c in camlClock__fun_2189 () at clock.ml:296
#11 0x00005601e564e123 in camlClock__loop_1139 () at clock.ml:266
#12 0x00005601e564c43d in camlClock__fun_2166 () at clock.ml:269
#13 0x00005601e5683a07 in camlTutils__process_729 () at tools/tutils.ml:175
#14 0x00005601e5983a25 in camlThread__fun_314 () at thread.ml:41
#15 0x00005601e5a7f400 in caml_start_program ()
#16 0x00005601e5a7573d in caml_callback_exn (closure=closure@entry=139910439183752, arg=<optimized out>, arg@entry=1) at callback.c:111
#17 0x00005601e5a55da0 in caml_thread_start (arg=0x7f3f680a6480) at st_stubs.c:538
#18 0x00007f3f84f45ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
#19 0x00007f3f81680def in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

I'm still investigating, sorry for the delay, this week has been crazy, but, clearly, memory allocations at `0x800` and `0x3` seems pretty suspicious... 🧐

**Comment by toots** [comment]
Here's the relevant code:
```c
CAMLprim value ocaml_lame_encode_buffer_float_ba(value l, value _bufl,
                                                 value _bufr) {
  CAMLparam3(l, _bufl, _bufr);
  CAMLlocal1(ret);
  lame_global_flags *lgf = Lame_val(l);
  struct caml_ba_array *bal = Caml_ba_array_val(_bufl);
  struct caml_ba_array *bar = Caml_ba_array_val(_bufr);
  int samples = bal->dim[0];
  if (bar->dim[0] != samples)
    caml_failwith("Invalid argument: buffers must be of same length");

  caml_release_runtime_system();

  unsigned char outbuf[LAME_MAXMP3BUFFER];
  int ans;

  ans = lame_encode_buffer_float(lgf, bal->data, bar->data, samples, outbuf,
                                 sizeof(outbuf));
```

I think the most reasonable explanation here is that the OCaml GC has moved the buffers around between the call to `caml_release_runtime_system` and the call to `lame_encode_buffer_float`, for instance for a memory compaction. This results in reads to `bal->data` and `bar->data` that are accessing the wrong memory area, returning faulty pointers.

If so, the solution is to extract those pointers _before_ releasing the OCaml runtime and GC:
```c
  struct caml_ba_array *bal = Caml_ba_array_val(_bufl);
  struct caml_ba_array *bar = Caml_ba_array_val(_bufr);
  const float *bufl = bal->data;
  const float *bufr = bar->data;
  int samples = bal->dim[0];
  if (bar->dim[0] != samples)
    caml_failwith("Invalid argument: buffers must be of same length");

  caml_release_runtime_system();

  unsigned char outbuf[LAME_MAXMP3BUFFER];
  int ans;

  ans = lame_encode_buffer_float(lgf, bufl, bufr, samples, outbuf,
                                 sizeof(outbuf));
```

I just pushed this change to https://github.com/savonet/ocaml-lame. @vitoyucepi any chance you could test with the latest `ocaml-lame` from there?

Again, thank you so much for your efforts with this.

**Comment by vitoyucepi** [comment]
@toots no problem.
Could you please create a github-ci build with new `deb` packages?
It's very convenient for me to use packages instead of building everything in place. So I use official builds from releases and ci.

**Comment by toots** [comment]
Great! There's a build coming up here: https://github.com/savonet/liquidsoap/actions/runs/1568125049

**Comment by vitoyucepi** [comment]
@toots do you want to retract `2.0.1` and then recreate it again, or this will be `2.0.2`?
As for me `2.0.2` is better, because it's easier to detect changes by automatic tools. Maybe some distro maintainers have already created packages with `2.0.1`.

**Comment by toots** [comment]
I'm not sure yet. It's an issue with one our dependent bindings so not directly with liquidsoap. Ideally I'd like to avoid having to do a release for each of these bugfixes. I might prolly bump the binary packages from the release page. That's how it would work out with the actual distribution packages.

Let me know if you can confirm that the bug is gone!

**Comment by vitoyucepi** [comment]
Well, there are no core dumps yet. Current uptime is _19 hours_.
Lets wait for some time, and then mark as solved. I think a _week_ will be enough.

As for the version. I have some bad experience with packages, that update old releases.
The problem is how to identify release rebuild for a package. Most of the monitoring tools will never check hashsums of the release.

In semver world it's good practice to bump patch for smallest changes. Also I've seen that broken releases get removed and then new versions tagged. 
For example https://github.com/composer/composer/compare/2.1.12...2.1.14.

However in docker world it's very common for official images to get rebuilds without version change, to track changes one can pin image hash in `Dockerfile`.

Distro maintainers use some kind of _release suffix_ for that purpose.

**Comment by toots** [comment]
> Well, there are no core dumps yet. Current uptime is _19 hours_. Lets wait for some time, and then mark as solved. I think a _week_ will be enough.
> 
> As for the version. I have some bad experience with packages, that update old releases. The problem is how to identify release rebuild for a package. Most of the monitoring tools will never check hashsums of the release.
> 
> In semver world it's good practice to bump patch for smallest changes. Also I've seen that broken releases get removed and then new versions tagged. For example [composer/composer@2.1.12...2.1.14](https://github.com/composer/composer/compare/2.1.12...2.1.14).
> 
> However in docker world it's very common for official images to get rebuilds without version change, to track changes one can pin image hash in `Dockerfile`.
> 
> Distro maintainers use some kind of _release suffix_ for that purpose.

I think for binary packages, I would bump the version, i.e. `2.0.1-2`. Docker tagging blows, frankly.

However, this is all theoretical. We're gonna do a `2.0.2` bugfix after all so, that solves our issues.

**Comment by EricVanGorkom** [comment]
Following the progress of this one, since I experience a similar lib3lame exception in one of my installs. Other ones (installed with exactly same OS and versions) works OK, this one don't. Only difference is that there is audio processing done in this install, not in the other ones. Would crash daily.

Installed the 2.0.1-working this morning, will add experiences with it in a few days.

**Comment by vitoyucepi** [comment]
@EricVanGorkom could you share a brief info about your environments.
- Do you build them by yourself or use prebuilt packages?
- What OS releases and package versions do you use?

Maybe this issue is kinda related to ocaml version.

**Comment by EricVanGorkom** [comment]
> @EricVanGorkom could you share a brief info about your environments.
> 
>     * Do you build them by yourself or use prebuilt packages?
> 
>     * What OS releases and package versions do you use?
> 
> 
> Maybe this issue is kinda related to ocaml version.

Sorry, was a bit too hasty (also saw that I used my GitHub workaccount instead of private one, but never mind) and forgot to provide those.

Using the prebuilt versions (for Focal AMD64) on a "vanilla" (i.e. only using them for audio with icecast and a few other small packages) Ubuntu 20.04-3, all with the latest packages. All machines have been installed and maintained with the same ansible scripts. Have 6 machines of those running with 2.0.1, all file input and streaming audio output (icecast). Input file is never a MP3, so liblame is only used for the output.

**Comment by vitoyucepi** [comment]
@EricVanGorkom so there's no difference between this instances?
All of them use the same version of liquidsoap from this [deb](https://github.com/savonet/liquidsoap/releases/download/v2.0.1/liquidsoap_2.0.1-ubuntu-focal-1_amd64.deb).
All of them encode mp3 streams.

I think that maybe all of them can crash some day due to this error.

**Comment by EricVanGorkom** [comment]
@vitoyucepi : They all have/had exactly the same package (since the package is copied and installed by ansible). The only one now different is the crashing instance because I've updated it this morning to the -working version.

Configurationfile wise, there is a difference, as mentioned, since the crashing one is doing some additional audio processing; that could be the main difference causing the crash. At first I suspected the processing part, but then I saw this report, and was wondering if it was related to my issues. 

Don't know if the other instances could crash; maybe the chance is just a lot less (because of the difference in configfile), but the other instances are running with 2.0.1 since the 10th of december and no issues. The one with issues is a new one, added the 10th and was crashing every day. So I don't know if earlier versions of liquidsoap would have the same issue.

**Comment by toots** [comment]
Hi all.

As far as I can see, the issue is linked to the `lame` binding specifically. It is a little tricky to detect wether or not you have a liquidsoap build with the fix since it is baked in depending on the version of `ocaml-lame` used at compile time.

* Currently, binary packages on the `2.0.0` and `2.0.1` release pages are subject to issue.
* Build via `opam` are if you are using official packages. You can in this case manually pin the latest `ocaml-lame`.

We have started a new release branch here: https://github.com/savonet/liquidsoap/pull/2104 All builds on this this branch should have the fix that I worked on, for instance here: https://github.com/savonet/liquidsoap/actions/runs/1579823311

I will release a new version of `ocaml-lame` to the `opam` package registry once @vitoyucepi confirms that the bug is not occurring for him.

**Comment by toots** [comment]
@vitoyucepi Have you experienced the bug since the fix?

**Comment by vitoyucepi** [comment]
@toots everything going fine. Uptime's 4 days.
So there were no segfaults yet.

**Comment by toots** [comment]
Awesome. Marking this as done. I will release a new `ocaml-lame` version with it today and it'll be part of the `2.0.2` release builds.
